hledger-lib/Hledger/Read/Common.hs

```diff
@@ -580,11 +581,14 @@ rawnumberp = do
         return (lastSep, fromMaybe [] digits)
 
     -- make sure we didn't leading part of mistyped number
-    notFollowedBy $ oneOf sepChars
+    notFollowedBy $ oneOf sepChars <|> (whitespaceChar >> digitChar)
```

Comment:
Space separators are blowing my mind just a little..

What's the reason you don't add space to sepChars ?
Does this accept other whitespace chars, like tab or line feed ?

Reply:
> Space separators are blowing my mind just a little..

I used space groups separator because that's one of the way most calculators do and that it doesn't mix like `,` with `.` which in some locales flipped. And it is hoarder to distinguish comma and dot in some fonts. Note that in some programming languages underscores are used to visually emphasize magniture like `1_000` which I believ is an attempt to get visually similar character to space.

> What's the reason you don't add space to sepChars ?

Space special and that's one ifts benefit. I think allowing amounts like `$ 1,000 45 @ 2 33 EUR` would be awful.
`sepChars` are generic and can be used either for groups or for decimals separator.

> Does this accept other whitespace chars, like tab or line feed ?

line-feed is a control character and doesn't fit Zs unicode category. Same for tabs.

See [wiki](https://en.wikipedia.org/wiki/Unicode_character_property)

> Includes the space, but not TAB, CR, or LF, which are Cc

Reply:
I think support for space digit group separator is nice. https://en.wikipedia.org/wiki/Decimal_mark#Digit_grouping indicates increasing support for it.

So it sounds like it accepts a (single) genuine space only. Wikipedia also suggests unicode thin space, but I'm not sure if we really want to go there - non-obvious unicode characters always end up causing problems somewhere..

Ah I see, we should probably split up sepChars some time.

Reply:
@simonmichael with this change we support [any kind of space within Zs category](https://www.compart.com/en/unicode/category/Zs) including "thin space" `U+2009`. I prefer to use non-breakinging space `U+00A0` in word processors (like LyX) to ensure that numbers are not teared apart.

Here we actually have 3 sets of characters:
- that can be used either as digit group separator or decimal part separator `sepChar`
- that can only be used as digit groups separator
- that can only be used as decimal part separator (currently empty)

Since last two groups always used in one place there is no much point in factoring out them.
The only problem is validation if number is finished correctly and not followed by something. With space you cannot just check that number is not followed by space because it might be used to separate number from commodity/comment/price. Thus we have that `whitespaceChar >> digitChar` instead of just `whitespaceChar`.

Reply:
Huh. So this change allows ascii space and *16* other kinds of unicode space.

Although we allow unicode in the journal, somehow this feels likely cause complications, hair-pulling, teeth-gnashing in future. I can also see the value of a non-breaking space in your use case. I guess I won't object if the list is for this.

Reply:
> Although we allow unicode in the journal, somehow this feels likely cause complications, hair-pulling, teeth-gnashing in future. I can also see the value of a non-breaking space in your use case. I guess I won't object if the list is for this.

I'm not sure I understand your concerns. Among all of the space characters ascii space is the most dangerous one because it used as delimiter of number and currency which at this moment can also contain non-ascii space characters because of `spacenonewline`.

I'd suggest to delegate most of the decision about interpreting characters to parser (`megaparsec`).

I think giving an unequal **treating** of whitespace characters actually is worse. Imagine that copying from posts from blogs and word processors that tend to flip those characters as they want resulting in journal where two spaces delimiter doesn't separate account from amount anymore because they were replaced with some long spaces.

Use of [categories](https://hackage.haskell.org/package/base-4.9.1.0/docs/Data-Char.html#t:GeneralCategory) will help to get rid of that burden at least for space separators (not that name of category pretty matching usage).

P.S. I've noticed that `spacenonewline` allows [vertical tab character (`\v`)](https://www.compart.com/en/unicode/U+000B) which appear like:
```shell
zsh% echo -e 'A\vB'
A
 B
```